moped-etl/arcgis/components_to_agol.py

```diff
@@ -115,17 +129,81 @@ def make_all_features(data):
                 attributes=component,
             )
             all_features["combined"].append(line_feature)
+
+            project_component_id = feature["attributes"]["project_component_id"]
+
+            # Find the elements in exploded_geometry which match the project_component_id
+            matching_elements = [
+                element
+                for element in exploded_geometry
+                if element["project_component_id"] == str(project_component_id)
+            ]
+
+            for component in matching_elements:
+
+                geometry_json = component.pop(
+                    "geometry"
+                )  # reminder, this is `pop()` on a dict, not an array
+
+                geometry = json.loads(
+                    geometry_json
+                )  # Decode JSON string into Python dictionary
+
+                if not geometry:
+                    continue
+                esri_geometry_key = get_esri_geometry_key(geometry)
+                feature = make_esri_feature(
+                    esri_geometry_key=esri_geometry_key,
+                    geometry=geometry,
+                    attributes=component,
+                )
+                feature["attributes"]["source_geometry_type"] = (
+                    "point" if esri_geometry_key == "points" else "line"
+                )
+                if not feature:
+                    continue
+
+                all_features["exploded"].append(feature)
         else:
             all_features["lines"].append(feature)
             all_features["combined"].append(feature)
 
     return all_features
 
 
+def transform_data(input_data):
+    if not input_data or len(input_data) < 2:
+        return []
+
+    field_names = input_data[0]
+    result = []
+
+    for row in input_data[1:]:
+        obj = {}
+        for field, value in zip(field_names, row):
+            if value == "NULL":
+                obj[field] = None
+            else:
+                try:
+                    obj[field] = int(value)
+                except ValueError:
+                    try:
+                        obj[field] = float(value)
+                    except ValueError:
+                        obj[field] = value
+        result.append(obj)
+
+    return result
+
+
```

Comment:
this function both changes the shape of the data coming back to an array of dictionaries, but also casts the data back into ints, floats and strings due to the "creative" use of the SQL endpoint. cc/ @mddilley

Reply:
My rational was that it would have been a 3 views deep, meaning any changes to the project list view mean dropping it, the 1st AGOL view, and this. Then make the change and have to build them all back up. That coupled with my mistaken belief that this was intended usage when I started got me here. The docs have an example of querying data out with using SQL via this part of the graphql-engine, but in retrospect, it's not meant for anything other than the most simple of queries. 

Anyway, it's all good - i'm removing the request for reviews - Mike gave me the same direction to go back and put a view in, so, I'm working on that today. 

Further down the road, I think a better solution might be to rework the AGOL view so that records don't represent components, but rather they represent single geometries, and then you can aggregate them into components much more easily than exploding pre-aggregated stuff back into single geometries. But -- another view will get us over the finish line and on to the next thing.

Thanks for you input -- you are undoubtedly correct, this is a bit to much to ask of that endpoint usage.

Reply:
Sounds good—thanks, Frank.

Reply:
This brought its own set of TILs, but this choice was 100% for the best. Thanks again for this request John. 🙏